### PR TITLE
Add Pins

### PR DESCRIPTION
**Pull request by jabbate19**

### CODE REVIEW

**Comment by jabbate19** [comment]
Screen Shot 2022-03-25 at 10 48 45 PM
Look, new pins. They're yellow. And come up first in the dir.

File: gallery/__init__.py [hunk 1]
```diff
@@ -618,6 +618,89 @@ def move_dir(dir_id: int, auth_dict: Optional[Dict[str, Any]] = None):
...
+            or auth_dict['is_organizer']):
+        return "Permission denied", 403
+    
+    dir_model.pinned = True
+
+    db.session.commit()
+
+    return "ok", 200
+
+@app.route("/api/file/unpin/<int:file_id>", methods=['POST'])
```
Comment (mxmeinhold) [inline gallery/__init__.py:663]: why unpin and not just methods=DELETE?

File: gallery/__init__.py [hunk 2]
```diff
@@ -876,21 +959,22 @@ def get_dir_children(dir_id: int) -> Any:
...
-def display_files(dir_id: int, internal: bool = False, auth_dict: Optional[Dict[str, Any]] = None):
+def display_files(dir_id: int, internal: bool = False, auth_dict: Optional[Dict[str, Any]] = None, pinned = False):
     gallery_lockdown = util.get_lockdown_status()
     if gallery_lockdown and (not auth_dict['is_eboard'] and not auth_dict['is_rtp']):
         abort(405)
 
-    file_list = [("File", f) for f in File.query.filter(File.parent == dir_id).all()]
-    dir_list = [("Directory", d) for d in Directory.query.filter(Directory.parent == dir_id).all()]
+    file_list = [("File", f) for f in File.query.filter(File.parent == dir_id, File.pinned == pinned).all()]
+    dir_list = [("Directory", d) for d in Directory.query.filter(Directory.parent == dir_id, Directory.pinned == pinned).all()]
```
Comment (mxmeinhold) [inline gallery/__init__.py:968]: Wait won't this result in not displaying pinned things if pinned isn't true?

File: gallery/__init__.py [hunk 3]
```diff
@@ -876,21 +959,22 @@ def get_dir_children(dir_id: int) -> Any:
...
 
     # Sort by name/title
     file_list.sort(key=lambda x: x[1].get_name())
     file_list.sort(key=lambda x: x[1].date_uploaded)
     dir_list.sort(key=lambda x: x[1].get_name())
     dir_list.sort(key=lambda x: x[1].date_uploaded)
 
-    ret_dict = dir_list + file_list
+    ret_dict = file_list + dir_list
```
Comment (mxmeinhold) [inline gallery/__init__.py:976]: Why swap these?

File: gallery/static/js/gallery.js [hunk 4]
```diff
@@ -315,6 +315,36 @@ function moveDir() {
     });
 }
 
+function pinFile(pinned, this_id) {
```
Comment (mxmeinhold) [inline gallery/static/js/gallery.js:318]: nit: call this toggleFilePin or something similar

File: migrations/versions/e2e671ede5a4_.py [hunk 5]
```diff
@@ -0,0 +1,28 @@
+"""empty message
```
Comment (mxmeinhold) [inline migrations/versions/e2e671ede5a4_.py:1]: :( no message

File: gallery/templates/view_dir.html [hunk 6]
```diff
@@ -32,7 +32,38 @@
             </ul>
         </div>
     </div>
-{% if children|length >= 1 %}
+{% if children|length + pin_children|length >= 1 %}
+    <div class="row" id="pin_child_list">
+    {% for child_type, child in pin_children %}
+        {% if not child.hidden or (auth_dict['is_eboard'] or auth_dict['is_rtp'])  or auth_dict['is_alumni'] %}
+            <div class="col-md-3 col-sm-4 col-xs-6">
+                <div class="panel panel-warning">
```
Comment (mxmeinhold) [inline gallery/templates/view_dir.html:40]: Would it be cleaner to do sorting in python and do the bg selection based on the pinned attribute of an item when rendering?

File: gallery/__init__.py (hunk 2 above)
Comment (jabbate19) [reply gallery/__init__.py:968]: Anything that is pinned will have pinned true.

File: gallery/__init__.py (hunk 3 above)
Comment (jabbate19) [reply gallery/__init__.py:976]: Whoops, I accidentally deleted line and thought I remembered the order.

File: gallery/templates/view_dir.html (hunk 6 above)
Comment (jabbate19) [reply gallery/templates/view_dir.html:40]: Hmmmm I will look into this

File: gallery/__init__.py [hunk 7]
```diff
@@ -618,6 +618,89 @@ def move_dir(dir_id: int, auth_dict: Optional[Dict[str, Any]] = None):
...
+    file_model = File.query.filter(File.id == file_id).first()
+
+    if file_model is None:
+        return "file not found", 404
+
+    assert auth_dict
+    if not (auth_dict['is_eboard']
+            or auth_dict['is_rtp']
+            or auth_dict['is_organizer']):
+        return "Permission denied", 403
```
[the remark this answers is not in the document]
Comment (jabbate19) [reply gallery/__init__.py:634]: I mimicked the other API sections for consistency

File: gallery/__init__.py [hunk 8]
```diff
@@ -618,6 +618,89 @@ def move_dir(dir_id: int, auth_dict: Optional[Dict[str, Any]] = None):
 
     return "ok", 200
 
+@app.route("/api/file/pin/<int:file_id>", methods=['POST'])
+@auth.oidc_auth('default')
+@gallery_auth
+def pin_file(file_id: int, auth_dict: Optional[Dict[str, Any]] = None):
+    file_model = File.query.filter(File.id == file_id).first()
```
Comment (mxmeinhold) [inline gallery/__init__.py:625]: You also need to check if the file is hidden I think.

File: gallery/__init__.py (hunk 2 above)
Comment (mxmeinhold) [reply gallery/__init__.py:968]: Can you add a doc comment to indicate that pinned means "only return pinned/unpinned"?

File: migrations/versions/e2e671ede5a4_.py [hunk 9]
```diff
@@ -0,0 +1,28 @@
+"""Add pinned to directories and files
```
Comment (mxmeinhold) [inline migrations/versions/e2e671ede5a4_.py:1]: You should also add a description to the file name

File: setup.cfg [hunk 10]
```diff
@@ -6,7 +6,7 @@ summary = Python Photo Gallery Written in Flask
 url = "https://github.com/ComputerScienceHouse/gallery"
 description-file = README.md
 license = MIT
-version = 2.2.0
+version = 2.3.0
```
Comment (mxmeinhold) [inline setup.cfg:9]: Why is this showing up in your PR?

File: gallery/static/js/gallery.js [hunk 11]
```diff
@@ -315,6 +315,36 @@ function moveDir() {
...
+    } else {
+        $.ajax({
+            type: "POST",
+            url: "/api/file/unpin/" + this_id,
+        });
+    }
+    location.reload();
+}
+
+function toggleDirPin(pinned, this_id) {
```
Comment (mxmeinhold) [inline gallery/static/js/gallery.js:333]: Honestly, why not just make a pin and unpin function rather than making it an arg, since it's not like any call uses a variable for the arg

File: setup.cfg (hunk 10 above)
Comment (jabbate19) [reply setup.cfg:9]: I need to rebase

File: gallery/__init__.py (hunk 1 above)
Comment (jabbate19) [reply gallery/__init__.py:665]: I don't see this is a deletion of anything, more of changing the state of the file/dir itself.

File: gallery/__init__.py [hunk 12]
```diff
@@ -123,7 +123,6 @@ def index():
     root_id = get_dir_tree(internal=True)
     return redirect("/view/dir/" + str(root_id['id']))
 
-
```
Comment (mxmeinhold) [inline gallery/__init__.py:126]: why was this removed? 2 lines between functions is standard in this file

File: gallery/__init__.py [hunk 13]
```diff
@@ -618,6 +617,96 @@ def move_dir(dir_id: int, auth_dict: Optional[Dict[str, Any]] = None):
 
     return "ok", 200
```
Comment (mxmeinhold) [inline gallery/__init__.py:619]:
```suggestion


```

File: gallery/__init__.py [hunk 14]
```diff
@@ -618,6 +617,96 @@ def move_dir(dir_id: int, auth_dict: Optional[Dict[str, Any]] = None):
...
+            or auth_dict['is_rtp']
+            or auth_dict['is_organizer']):
+        return "Permission denied", 403
+    
+    file_model.pinned = True
+
+    db.session.commit()
+
+    return "ok", 200
+
```
Comment (mxmeinhold) [inline gallery/__init__.py:643]:
```suggestion


```

File: gallery/__init__.py [hunk 15]
```diff
@@ -618,6 +617,96 @@ def move_dir(dir_id: int, auth_dict: Optional[Dict[str, Any]] = None):
...
+            or auth_dict['is_rtp']
+            or auth_dict['is_organizer']):
+        return "Permission denied", 403
+    
+    dir_model.pinned = True
+
+    db.session.commit()
+
+    return "ok", 200
+
```
Comment (mxmeinhold) [inline gallery/__init__.py:664]:
```suggestion


```

File: gallery/__init__.py [hunk 16]
```diff
@@ -618,6 +618,89 @@ def move_dir(dir_id: int, auth_dict: Optional[Dict[str, Any]] = None):
...
+    dir_model = Directory.query.filter(Directory.id == dir_id).first()
+
+    if dir_model is None:
+        return "directory not found", 404
+
+    assert auth_dict
+    if not (auth_dict['is_eboard']
+            or auth_dict['is_rtp']
+            or auth_dict['is_organizer']):
+        return "Permission denied", 403
```
[the remark this answers is not in the document]
Comment (mxmeinhold) [reply gallery/__init__.py:657]: Hidden check?

File: gallery/__init__.py (hunk 1 above)
Comment (mxmeinhold) [reply gallery/__init__.py:665]: 🤷 I kinda see it as `delete a pin` but both work.

File: gallery/__init__.py [hunk 17]
```diff
@@ -618,6 +617,96 @@ def move_dir(dir_id: int, auth_dict: Optional[Dict[str, Any]] = None):
...
+    
+    file_model.pinned = False
+
+    db.session.commit()
+
+    return "ok", 200
+
```
Comment (mxmeinhold) [inline gallery/__init__.py:689]:
```suggestion


```

File: gallery/__init__.py [hunk 18]
```diff
@@ -618,6 +617,96 @@ def move_dir(dir_id: int, auth_dict: Optional[Dict[str, Any]] = None):
...
+    if not (auth_dict['is_eboard']
+            or auth_dict['is_rtp']
+            or auth_dict['is_organizer']):
+        return "Permission denied", 403
+    
+    dir_model.pinned = False
+
+    db.session.commit()
+
+    return "ok", 200
```
Comment (mxmeinhold) [inline gallery/__init__.py:710]:
```suggestion


```

File: gallery/__init__.py [hunk 19]
```diff
@@ -618,6 +617,96 @@ def move_dir(dir_id: int, auth_dict: Optional[Dict[str, Any]] = None):
...
+
+    if dir_model is None:
+        return "directory not found", 404
+
+    assert auth_dict
+    if not (auth_dict['is_eboard']
+            or auth_dict['is_rtp']
+            or auth_dict['is_organizer']):
+        return "Permission denied", 403
+
```
Comment (mxmeinhold) [inline gallery/__init__.py:704]: does this need a hidden check?

File: gallery/__init__.py [hunk 20]
```diff
@@ -876,13 +965,21 @@ def get_dir_children(dir_id: int) -> Any:
 @app.route("/api/directory/get/<int:dir_id>")
 @auth.oidc_auth('default')
 @gallery_auth
-def display_files(dir_id: int, internal: bool = False, auth_dict: Optional[Dict[str, Any]] = None):
+def display_files(dir_id: int, internal: bool = False, auth_dict: Optional[Dict[str, Any]] = None, pinned = False):
+    """
+        Grab all children of the given directory, sort, and return as list
+
+        dir_id - ID of the directory to find children of
```
Comment (mxmeinhold) [inline gallery/__init__.py:972]:
```suggestion
        dir_id: ID of the directory to find children of
```
for consistency

File: gallery/__init__.py [hunk 21]
```diff
@@ -876,13 +965,21 @@ def get_dir_children(dir_id: int) -> Any:
...
+        internal: If the call is being made internally, return python object instead of converting to JSON
+        auth_dict: Used for checking if user is E-Board or RTP
+        pinned: Only return pinned or unpinned children
```
Comment (mxmeinhold) [inline gallery/__init__.py:975]: nit: this is ambiguous `if true, return only pinned children, otherwise, return only unpinned children` maybe?

File: gallery/templates/view_dir.html [hunk 22]
```diff
@@ -83,6 +88,13 @@ <h2>This album is empty.</h2>
         <a href="#!" class="btn btn-primary" onclick="editDirDescription()">Edit</a>
         {% if auth_dict['can_edit'] or auth_dict['is_organizer'] %}
         <a href="#!" class="btn btn-warning" onclick="moveDir()">Move</a>
+        {% if directory.parent != None %}
+            {% if directory.pinned %}
+            <a href="#!" class="btn btn-warning" onclick="unpinDir({{ directory.id }})">Unpin</a>
+            {% else %}
+            <a href="#!" class="btn btn-warning" onclick="pinDir({{ directory.id }})">Pin</a>
+            {% endif %}
+        {% endif %}
```
Comment (mxmeinhold) [inline gallery/templates/view_dir.html:97]: this will allow uploaders to see the pin buttons, no?

Comment (mxmeinhold) [reply gallery/templates/view_dir.html:97]: also
curious: why directory.parent != None?

File: gallery/__init__.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (jabbate19) [reply gallery/__init__.py:657]: Directories cannot be hidden, only files

File: gallery/__init__.py (hunk 19 above)
Comment (jabbate19) [reply gallery/__init__.py:704]: Directories cannot be hidden, only files

File: gallery/templates/view_dir.html (hunk 22 above)
Comment (jabbate19) [reply gallery/templates/view_dir.html:97]: Don't want to have the pin button on the root directory, it's pointless

**Review by mxmeinhold — APPROVED** [review]
